### ISSUE TITLE
output formatting: -x-1 has redundant parentheses: (-x)-1

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#2102]
Imported from SourceForge on 2024-07-05 18:39:27
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2023-01-27 23:38:19
Original: https://sourceforge.net/p/maxima/bugs/4090

---

Both in 2d and 1d output, the expression `-x-1` is formatted as `(-x)-1`. The parentheses are unnecessary and should not be printed.

This bug appears to affect only the first element of a sum:
~~~
(-b)-a 
(-c)-a 
(-c)-b 
(-c)+b+a 
(-c)+b-a 
(-c)-b+a 
(-c)-b-a 
-a 
a-b 
a-c 
b-a 
b-c 
c+b-a 
c-a 
c-b 
c-b+a 
c-b-a 
~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:39:28
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-01-28 01:12:55
Original: https://sourceforge.net/p/maxima/bugs/4090/#120e

---

As you know, the display of parentheses has been adjusted over the years -- looks like twice it was adjusted in the direction of more parentheses. See commits 8e6b48 and 592689.

`a + b - c` is reformatted to `((MPLUS) ((MMINUS) $C) $B $A)` , and that gets parentheses because LBP and RBP of MMINUS are the same as for MPLUS. 

Instead of changing the policy for assigning parentheses in DIMENSION, I experimented with increasing RBP and LBP for MMINUS. That makes some sense, right? because MMINUS is more like MTIMES , being an implicit multiplication by -1. Increasing LBP and RBP for MMINUS by 1 has the expected effect of suppressing parentheses in the cases above. `run_testsuite` reports a few failed tests which are looking for specific formatted output. A couple of those are trivial and obviously okay as revised; problem 653 in rtest16 is more complex, and I think it's okay, but I'll take another look.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:39:32
Created by **[rtoy](https://sourceforge.net/u/rtoy/)** on 2023-01-28 01:18:10
Original: https://sourceforge.net/p/maxima/bugs/4090/#e121

---

This has always annoyed me, especially for things like `-1-2*%i` being printed as `(-2*%i)-1`.

But after all these years, I've kind of gotten us to it.  But I wouldn't object if it got changed.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:39:35
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-01-28 18:13:13
Original: https://sourceforge.net/p/maxima/bugs/4090/#f662

---

I'm going to bump the binding powers for MMINUS (and MUNARYMINUS while I'm at it). Towards that end, here's a lit of operators with similar binding powers -- I'm guessing these are the ones for which changing parentheses might possibly yield different results, so maybe they should be verified in combination with MMINUS.
```
(%i1) :lisp (do-symbols (s :maxima) (if (and (get s 'rbp) (>= (get s 'rbp) 100) (<= (get s 'rbp) 120)) (print (list s (get s 'rbp)))))

(MPLUS 100) 
(%AT 105) 
(MQUOTIENT 120) 
(%PRODUCT 115) 
(MTIMES 120) 
(RAT 120) 
(MMINUS 100) 
(MUNARYMINUS 100) 
(%INTEGRATE 115) 
(%LIMIT 110) 
(%LSUM 110) 
($/ 120) 
(%SUM 110) 
NIL
(%i1) :lisp (do-symbols (s :maxima) (if (and (get s 'lbp) (>= (get s 'lbp) 100) (<= (get s 'lbp) 120)) (print (list s (get s 'lbp)))))

(MPLUS 100) 
(%AT 105) 
(MQUOTIENT 120) 
(MTIMES 120) 
(RAT 120) 
($- 100) 
(MMINUS 100) 
($* 120) 
(%LIMIT 110) 
($/ 120) 
($+ 100) 
NIL
```

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:39:39
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-02-04 17:26:15
Original: https://sourceforge.net/p/maxima/bugs/4090/#fc7c

---

I've pushed commit [ccf37eb6] as a proposed bug fix. The only functional change is to increase right-binding power of MMINUS from 100 to 134. This changes the display of MMINUS expressions but does not change parsing (which looks at rbp of `$-` instead). 

That commit doesn't totally close the book, however. There is still some possibly-unexpected redundant parentheses which appear in 1-d output (while not appearing in 2-d output). E.g.
```
(%i12) -2*%i-1;
(%o12)                     - 2 %i - 1
(%i13) string (%);
(%o13)                      -(2*%i)-1
```
The apparently-unneeded parentheses are due to NFORMAT reorganizing MTIMES applied to MMINUS to MMINUS applied to MTIMES. Tracing NFORMAT shows, in part,
```
2. Trace: (NFORMAT '((MTIMES SIMP) -2 $%I))
3. Trace: (NFORMAT '-2)
3. Trace: NFORMAT ==> ((MMINUS) 2)
3. Trace: (NFORMAT '2)
3. Trace: NFORMAT ==> 2
3. Trace: (NFORMAT '$%I)
3. Trace: NFORMAT ==> $%I
2. Trace: NFORMAT ==> ((MMINUS) ((MTIMES) 2 $%I))
```
Then displaying parentheses on the reorganized expression is correct, since otherwise the parser would not yield MMINUS as the top-level operator.

If we don't like that, I think we should talk about NFORMAT at this point.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:39:42
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-02-09 06:28:15
Original: https://sourceforge.net/p/maxima/bugs/4090/#5262

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:39:46
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-02-09 06:28:15
Original: https://sourceforge.net/p/maxima/bugs/4090/#68a3

---

I'm closing this ticket as fixed by the commit I mentioned before, [ ccf37eb6 ], as I'm satisfied at this point that the parentheses have been improved. If anyone disagrees, no problem, we can reopen this ticket or open a new one and keep working on it.
